docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1056 @@
+- [Introduction to vectorization with Vector128 and Vector256](#introduction-to-vectorization-with-vector128-and-vector256)
+  * [Code structure](#code-structure)
+    + [Testing](#testing)
+    + [Benchmarking](#benchmarking)
+      - [Custom config](#custom-config)
+      - [Memory alignment](#memory-alignment)
+        * [Enforcing memory alignment](#enforcing-memory-alignment)
+        * [Memory randomization](#memory-randomization)
+  * [Loops](#loops)
+    + [Scalar remainder handling](#scalar-remainder-handling)
+    + [Vectorized remainder handling](#vectorized-remainder-handling)
+    + [Access violation testing](#access-violation-av-testing)
+  * [Loading and storing vectors](#loading-and-storing-vectors)
+    + [Loading](#loading)
+    + [Storing](#storing)
+    + [Casting](#casting)
+  * [Mindset](#mindset)
+    + [Edge cases](#edge-cases)
+    + [Scalar solution](#scalar-solution)
+    + [Vectorized solution](#vectorized-solution)
+  * [Toolchain](#toolchain)
+    + [Creation](#creation)
+    + [Bit operations](#bit-operations)
+    + [Equality](#equality)
+    + [Comparison](#comparison)
+    + [Math](#math)
+    + [Conversion](#conversion)
+    + [Widening and Narrowing](#widening-and-narrowing)
+    + [Shuffle](#shuffle)
+  * [Summary](#summary)
+    + [Best practices](#best-practices)
+
+TL;DR: Go to [Summary](#summary)
+
+# Introduction to vectorization with Vector128 and Vector256
+
+Vectorization is the art of converting an algorithm from operating on a single value at a time to operating on a set of values (vector). It can greatly improve performance at a cost of increased code complexity.
+
+In recent releases, .NET has introduced many new APIs for vectorization. The vast majority of them are hardware specific, so they require users to provide an implementation per processor architecture (x64 and/or arm64), with the option of using the most optimal instructions for hardware that is executing the code.
+
+.NET 7 introduced a set of new APIs for `Vector128<T>` and `Vector256<T>` for writing hardware-agnostic, cross platform vectorized code. The purpose of this document is to introduce you to the new APIs and provide a set of best practices.
```

Comment:
We also introduced `Vector64<T>` at the same time. `Vector512<T>` is being introduced in .NET 8.

It might also be worth a small addendum that while all these have "software fallbacks" not all of them may be vectorized on all platforms. `Vector64<T>` is only accelerated on Arm64 today and `Vector256<T>` only on "newer" x86/x64.

`Vector128<T>` is the "common denominator" across all platforms that support vectorization (and this is expected to always be the case).